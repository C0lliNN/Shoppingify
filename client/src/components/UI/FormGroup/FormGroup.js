import React from 'react';
import styled from 'styled-components';
import {
  BORDER_RADIUS_1,
  COLORS,
  FONT_SIZE_2,
  FONT_FAMILY,
} from '../../../helpers/style-constants';
import PropTypes from 'prop-types';

const FormGroup = styled.div`
  display: flex;
  align-items: center;
  justify-content: start;
  flex-direction: column;
  margin-bottom: 10px;
`;

const formControlStyles = `
  display: block;
  order: 2;
  width: 100%;
  box-sizing: border-box;
  padding: 20px 12px;
  border: 2px solid ${COLORS.gray_2};
  border-radius: ${BORDER_RADIUS_1};
  outline: none;
  font-size: ${FONT_SIZE_2};
  font-family: ${FONT_FAMILY};
  font-weight: 500;
  color: ${COLORS.black_2};
  &:focus {
    border-color: ${COLORS.primary};
    box-shadow: 0px 0px 1px rgba(0, 0, 0, 0.3);
    +label {
      color: ${COLORS.primary}
    }
  }
  &::placeholder {
    color: ${COLORS.gray_2};
  };
`;
const Input = styled.input`
  ${formControlStyles}
`;

const Textarea = styled.textarea`
  ${formControlStyles}
`;

const Select = styled.select`
  ${formControlStyles}
`;

const Label = styled.label`
  display: block;
  width: 100%;
  color: ${COLORS.black_2};
  font-weight: 600;
  margin-bottom: 5px;
  font-size: ${FONT_SIZE_2};
`;

function Wrapper(props) {
  const children = [];

  if (Array.isArray(props.children)) {
    const labelIndex = props.children.findIndex(
      (element) => element.type.displayName === 'styled.label'
    );

    for (let i = 0; i < props.children.length; i++) {
      if (i !== labelIndex) {
        children.push(props.children[i]);
      }
    }

    children.push(props.children[labelIndex]);
  } else {
    children.push(props.children);
  }

  return <FormGroup>{children}</FormGroup>;
}

Wrapper.Input = Input;
Wrapper.Textarea = Textarea;
Wrapper.Select = Select;
Wrapper.Label = Label;

Wrapper.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default Wrapper;
