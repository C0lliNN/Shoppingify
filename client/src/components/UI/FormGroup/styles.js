import styled from 'styled-components';
import * as variables from '../../../helpers/style-constants';

export const StyledFromGroup = styled.div`
  display: flex;
  align-items: center;
  justify-content: start;
  flex-direction: column;
  margin-bottom: 25px;

  & span {
    order: 3;
  }
`;

export const formControlStyles = `
  display: block;
  order: 2;
  width: 100%;
  box-sizing: border-box;
  padding: 20px 12px;
  border: 2px solid ${variables.COLORS.gray_2};
  border-radius: ${variables.BORDER_RADIUS_1};
  outline: none;
  font-size: ${variables.FONT_SIZE_2};
  font-family: ${variables.FONT_FAMILY};
  font-weight: 500;
  color: ${variables.COLORS.black_2};
  &:focus {
    border-color: ${variables.COLORS.primary};
    box-shadow: 0px 0px 1px rgba(0, 0, 0, 0.3);
    +label {
      color: ${variables.COLORS.primary}
    }
  }
  &::placeholder {
    color: ${variables.COLORS.gray_2};
  };
`;
export const Input = styled.input`
  ${formControlStyles}
`;

export const Textarea = styled.textarea`
  ${formControlStyles}
`;

export const Select = styled.select`
  ${formControlStyles}
`;

export const Label = styled.label`
  display: block;
  width: 100%;
  color: ${variables.COLORS.black_2};
  font-weight: 600;
  margin-bottom: 5px;
  font-size: ${variables.FONT_SIZE_2};
`;
