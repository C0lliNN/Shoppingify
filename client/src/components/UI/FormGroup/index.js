import PropTypes from 'prop-types';
import React from 'react';
import { StyledFromGroup, Input, Textarea, Select, Label } from './styles';

function FormGroup(props) {
  const children = [];

  if (Array.isArray(props.children)) {
    const labelIndex = props.children.findIndex(
      (element) => element.type.displayName === 'styled.label'
    );

    for (let i = 0; i < props.children.length; i++) {
      if (i !== labelIndex) {
        children.push(props.children[i]);
      }
    }

    children.push(props.children[labelIndex]);
  } else {
    children.push(props.children);
  }

  return <StyledFromGroup>{children}</StyledFromGroup>;
}

FormGroup.Input = Input;
FormGroup.Textarea = Textarea;
FormGroup.Select = Select;
FormGroup.Label = Label;

FormGroup.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

export default FormGroup;
