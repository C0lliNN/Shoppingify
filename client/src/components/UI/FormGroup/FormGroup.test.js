import React from 'react';
import { render } from '@testing-library/react';
import FormGroup from './FormGroup';
import 'jest-styled-components';

describe('<FormGroup/>', () => {
  let errorMessage = null;
  beforeEach(() => {
    console.error = jest.fn((message) => (errorMessage = message));
  });
  it('should console.error if no children elements are provided', () => {
    render(<FormGroup></FormGroup>);

    expect(console.error).toHaveBeenCalled();
    expect(errorMessage).toMatch(/.*children.*required.*/);
  });
  it('should render a input correctly if no label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Input placeholder="test" />
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
  it('should render a textarea correctly if no label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Textarea placeholder="test"></FormGroup.Textarea>
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
  it('should render a select correctly if no label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Select>
          <option value="1">Test</option>
        </FormGroup.Select>
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
  it('should render a input correctly if a label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Label>Test</FormGroup.Label>
        <FormGroup.Input placeholder="test" />
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
  it('should render a textarea correctly if a label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Label>Test</FormGroup.Label>
        <FormGroup.Textarea placeholder="test"></FormGroup.Textarea>
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
  it('should render a select correctly if a label is provided', () => {
    const { container } = render(
      <FormGroup>
        <FormGroup.Label>Test</FormGroup.Label>
        <FormGroup.Select>
          <option value="1">Test</option>
        </FormGroup.Select>
      </FormGroup>
    );

    expect(container).toMatchSnapshot();
  });
});
