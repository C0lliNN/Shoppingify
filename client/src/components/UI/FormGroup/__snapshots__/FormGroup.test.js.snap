// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormGroup/> should render a input correctly if a label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <input
      class="sc-AxirZ kKTTBK"
      placeholder="test"
    />
    <label
      class="sc-AxhUy XuytY"
    >
      Test
    </label>
  </div>
</div>
`;

exports[`<FormGroup/> should render a input correctly if no label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <input
      class="sc-AxirZ kKTTBK"
      placeholder="test"
    />
  </div>
</div>
`;

exports[`<FormGroup/> should render a select correctly if a label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <select
      class="sc-AxhCb SaXZS"
    >
      <option
        value="1"
      >
        Test
      </option>
    </select>
    <label
      class="sc-AxhUy XuytY"
    >
      Test
    </label>
  </div>
</div>
`;

exports[`<FormGroup/> should render a select correctly if no label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <select
      class="sc-AxhCb SaXZS"
    >
      <option
        value="1"
      >
        Test
      </option>
    </select>
  </div>
</div>
`;

exports[`<FormGroup/> should render a textarea correctly if a label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <textarea
      class="sc-AxiKw jdYDTq"
      placeholder="test"
    />
    <label
      class="sc-AxhUy XuytY"
    >
      Test
    </label>
  </div>
</div>
`;

exports[`<FormGroup/> should render a textarea correctly if no label is provided 1`] = `
<div>
  <div
    class="sc-AxjAm ediThQ"
  >
    <textarea
      class="sc-AxiKw jdYDTq"
      placeholder="test"
    />
  </div>
</div>
`;
